presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.facebook.presto.memory.LocalMemoryManager;
+import com.facebook.presto.memory.MemoryPool;
+import com.facebook.presto.memory.QueryContext;
+import com.facebook.presto.memory.TraversingQueryContextVisitor;
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Ordering;
+import io.airlift.log.Logger;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class MemoryRevokingScheduler
+{
+    private static final Logger log = Logger.get(MemoryRevokingScheduler.class);
+
+    private static final Ordering<SqlTask> ORDER_BY_CREATE_TIME = Ordering.natural().onResultOf(task -> task.getTaskInfo().getStats().getCreateTime());
```

Comment:
Hmm. It should not starve as there should be some progress always. But I agree they could be slowed down considerably in case they are spilling constantly.
Generally ordering here was introduced mostly to make all of the nodes spill similar set of queries. So the global (cluster wide) set of queries that are spilling memory at the moment (hence blocked) is smaller. Using reversed ordering (newest first) seems similarly good for that purpose. But honestly I am not sure what are the runtime characteristics of this code without testing it in production under considerable load. And I am pretty sure that mechanism will need to be extended over time. 
E.g. we are thinking that when spilling starts for a query on one node, it is reasonable to schedule spilling of that query on other nodes ahead of time. 
Also it may be reasonable to take into consideration how much revocable memory is used by query when selecting the ones to spill. So we schedule spilling for fewer but heavier rather than more but lighter queries.

We can change that but I do not believe we have data right now to pick a solution which will surly be better. Any suggestions?